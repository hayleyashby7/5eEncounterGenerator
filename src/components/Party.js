import React from 'react';

const Party = ({ onSubmit }) => {
	return (
		<div>
			<form onSubmit={onSubmit}>
				<label htmlFor='char-num'>Characters (Max 10)</label>
				<input type='number' id='char-num' min='1' max='10' />
				<label htmlFor='level'>Level (1-20)</label>
				<input type='number' id='level' min='1' max='20' />
				<button name='partySubmit' type='submit'>
					Submit
				</button>
			</form>
		</div>
	);
};

export default Party;
