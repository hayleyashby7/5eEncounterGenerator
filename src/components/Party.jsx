import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import Difficulty from '../utils/difficulty';
import { Input, Select } from './Form';

function Party({ saveData }) {
    const {
        register,
        formState: { errors },
        handleSubmit,
    } = useForm({ mode: 'onBlur' });

    const onSubmit = (data) => {
        saveData(data);
    };

    const difficultyArray = [
        { value: null, label: 'Select a difficulty' },
        { value: Difficulty.Easy, label: 'Easy' },
        { value: Difficulty.Medium, label: 'Medium' },
        { value: Difficulty.Hard, label: 'Hard' },
        { value: Difficulty.Deadly, label: 'Deadly' },
    ];

    return (
        <form onSubmit={handleSubmit(onSubmit)} className='flex flex-col flex-wrap lg:flex-row'>
            <div className='flex flex-1 flex-col'>
                <Input
                    label='Number of characters'
                    name='numCharacters'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 10,
                        message: 'Max 10',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>

            <div className='flex flex-1 flex-col'>
                <Input
                    label='Level'
                    name='level'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 20,
                        message: 'Max 20',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col'>
                <Select
                    label='Difficulty'
                    name='difficulty'
                    options={difficultyArray}
                    register={register}
                    required
                    validate={(value) =>
                        Difficulty.difficultyType(value) || 'Must select a difficulty'
                    }
                />
                <ErrorMessage
                    errors={errors}
                    name='difficulty'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col'>
                <input name='partySubmit' type='submit' />
            </div>
        </form>
    );
}

Party.propTypes = { saveData: PropTypes.func };
Party.defaultProps = { saveData: () => {} };

export default Party;
