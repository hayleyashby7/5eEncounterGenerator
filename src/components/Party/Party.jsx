import React from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import Difficulty from '../../utils/difficulty';
import Input from '../Form/Input';
import Select from '../Form/Select';
import Button from '../Form/Button';

function Party({ saveData }) {
    const {
        register,
        formState: { errors },
        handleSubmit,
    } = useForm({ mode: 'onBlur' });

    const onSubmit = (data) => {
        saveData(data);
    };

    const difficultyArray = [
        { value: null, label: 'Select a difficulty' },
        { value: Difficulty.Easy, label: 'Easy' },
        { value: Difficulty.Medium, label: 'Medium' },
        { value: Difficulty.Hard, label: 'Hard' },
        { value: Difficulty.Deadly, label: 'Deadly' },
    ];

    return (
        <form
            onSubmit={handleSubmit(onSubmit)}
            className='my-5 flex flex-col flex-wrap justify-end gap-4 rounded-lg bg-stone-400 p-5 text-red-950 lg:flex-row lg:justify-between'
        >
            <div className='flex flex-1 flex-col'>
                <Input
                    label='Number of characters'
                    name='numCharacters'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 10,
                        message: 'Max 10',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='numCharacters'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>

            <div className='flex flex-1 flex-col'>
                <Input
                    label='Level'
                    name='level'
                    type='number'
                    register={register}
                    required='Required'
                    valueAsNumber='Must be a number'
                    max={{
                        value: 20,
                        message: 'Max 20',
                    }}
                    min={{
                        value: 1,
                        message: 'Min 1',
                    }}
                />
                <ErrorMessage
                    errors={errors}
                    name='level'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col'>
                <Select
                    label='Difficulty'
                    name='difficulty'
                    options={difficultyArray}
                    register={register}
                    required
                    validate={(value) =>
                        Difficulty.difficultyType(value) || 'Must select a difficulty'
                    }
                />
                <ErrorMessage
                    errors={errors}
                    name='difficulty'
                    render={({ message }) => <p>{message}</p>}
                />
            </div>
            <div className='flex flex-1 flex-col justify-end'>
                <Button label='Generate' type='submit' name='generate' />
            </div>
        </form>
    );
}

Party.propTypes = { saveData: PropTypes.func };
Party.defaultProps = { saveData: () => {} };

export default Party;
