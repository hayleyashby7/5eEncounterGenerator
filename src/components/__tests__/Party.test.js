import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Party from '../Party';

describe('Party component', () => {
	it('renders the party composition form', () => {
		// Arrange
		const { container } = render(<Party />);

		//Assert
		expect(container).toMatchSnapshot();
	});

	it('calls onSubmit when Submit button is pressed', async () => {
		// Arrange
		const onSubmitMock = jest.fn();
		const user = userEvent.setup();
		render(<Party onSubmit={onSubmitMock} />);
		const button = screen.getByRole('button', { name: 'Submit' });

		// Act
		await user.click(button);

		//Assert
		expect(onSubmitMock).toHaveBeenCalledTimes(1);
	});

	it('form validation rejects number of characters outside of 1-10 values', async () => {
		// Arrange
		const onSubmitMock = jest.fn();
		const user = userEvent.setup();
		render(<Party onSubmit={onSubmitMock} />);
		const button = screen.getByRole('button', { name: 'Submit' });
		const input = screen.getByLabelText('Characters (Max 10)');

		// Act
		await user.type(input, '25');
		await user.click(button);

		//Asser
		expect(input).toBeInvalid();
	});

	it('form validation  rejects character level outside of 1-20 values', async () => {
		// Arrange
		const onSubmitMock = jest.fn();
		const user = userEvent.setup();
		render(<Party onSubmit={onSubmitMock} />);
		const button = screen.getByRole('button', { name: 'Submit' });
		const input = screen.getByLabelText('Level (1-20)');

		// Act
		await user.type(input, '68');
		await user.click(button);

		//Asser
		expect(input).toBeInvalid();
	});
});
