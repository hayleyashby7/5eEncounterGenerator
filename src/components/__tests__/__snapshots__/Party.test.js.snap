// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Party component renders the party composition form 1`] = `
<div>
  <div>
    <form>
      <label
        for="char-num"
      >
        Characters (Max 10)
      </label>
      <input
        id="char-num"
        max="10"
        min="1"
        type="number"
      />
      <label
        for="level"
      >
        Level (1-20)
      </label>
      <input
        id="level"
        max="20"
        min="1"
        type="number"
      />
      <button
        name="partySubmit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
